### PR TITLE
[SYCL][ESIMD] Add support for accessors to atomic_update API

### PR DESCRIPTION
**Pull request by fineg74**

### CODE REVIEW

File: sycl/include/sycl/ext/intel/esimd/memory.hpp [hunk 1]
```diff
@@ -1215,6 +1215,323 @@ atomic_update(Tx *p, Toffset offset, simd<Tx, N> src0, simd<Tx, N> src1,
                                   mask);
 }
 
+/// @anchor acessor_atomic_update1
```
Comment (sarnex) [inline sycl/include/sycl/ext/intel/esimd/memory.hpp:1218]:
```suggestion
/// @anchor accessor_atomic_update1
```

File: sycl/include/sycl/ext/intel/esimd/memory.hpp [hunk 2]
```diff
@@ -1215,6 +1215,323 @@ atomic_update(Tx *p, Toffset offset, simd<Tx, N> src0, simd<Tx, N> src1,
...
+__ESIMD_API std::enable_if_t<std::is_integral_v<Toffset> &&
+                                 !std::is_pointer<AccessorTy>::value,
+                             simd<Tx, N>>
+atomic_update(AccessorTy acc, Toffset offset, simd_mask<N> mask = 1) {
+  uint32_t loc_offset = offset;
+  return atomic_update<Op, Tx, N>(acc, simd<Toffset, N>(loc_offset), mask);
+}
+
+/// @anchor accessor_atomic_update2
+/// Atomically updates \c N memory locations represented by anaccessor and
```
Comment (sarnex) [inline sycl/include/sycl/ext/intel/esimd/memory.hpp:1432]:
```suggestion
/// Atomically updates \c N memory locations represented by an accessor and
```

File: sycl/test-e2e/ESIMD/dword_atomic_accessor_cmpxchg.cpp [hunk 3]
```diff
@@ -0,0 +1,27 @@
+//==------ dword_atomic_accessor_smoke.cpp  - DPC++ ESIMD on-device test --==//
```
Comment (sarnex) [inline sycl/test-e2e/ESIMD/dword_atomic_accessor_cmpxchg.cpp:1]: nit: comment in filename is wrong

File: sycl/test-e2e/ESIMD/dword_atomic_accessor_smoke.cpp [hunk 4]
```diff
@@ -0,0 +1,22 @@
+//==------ dword_atomic_accessor__smoke.cpp  - DPC++ ESIMD on-device test --==//
```
Comment (sarnex) [inline sycl/test-e2e/ESIMD/dword_atomic_accessor_smoke.cpp:1]:
```suggestion
//==------ dword_atomic_accessor_smoke.cpp  - DPC++ ESIMD on-device test --==//
```

File: sycl/test-e2e/ESIMD/dword_atomic_accessor_smoke_stateless.cpp [hunk 5]
```diff
@@ -0,0 +1,24 @@
...
+// enforced, i.e. accessor based accesses are automatically converted to
+// stateless accesses.
+//===----------------------------------------------------------------------===//
+// REQUIRES: gpu
+// UNSUPPORTED: gpu-intel-gen9 && windows
+// UNSUPPORTED: cuda || hip
+// RUN: %clangxx -fsycl -fsycl-esimd-force-stateless-mem %s -o %t.out
+// RUN: %GPU_RUN_PLACEHOLDER %t.out
+// TODO add support for __esimd_dword_atomic on esimd_emulator
+// XFAIL: esimd_emulator
```
Comment (sarnex) [inline sycl/test-e2e/ESIMD/dword_atomic_accessor_smoke_stateless.cpp:18]: if we are sure it always fails these is fine, if it could ever pass we should use `UNSUPPORTED`

File: sycl/include/sycl/ext/intel/esimd/memory.hpp [hunk 6]
```diff
@@ -1215,6 +1215,323 @@ atomic_update(Tx *p, Toffset offset, simd<Tx, N> src0, simd<Tx, N> src1,
...
+}
+
+/// A variation of \c atomic_update API with \c offset represented as
+/// scalar object.
+///
+/// @tparam Op The atomic operation - can be one of the following:
+/// \c atomic_op::add, \c atomic_op::sub, \c atomic_op::min, \c atomic_op::max,
+/// \c atomic_op::xchg, \c atomic_op::bit_and, \c atomic_op::bit_or,
+/// \c atomic_op::bit_xor, \c atomic_op::minsint, \c atomic_op::maxsint,
+/// \c atomic_op::fmax, \c atomic_op::fmin \c atomic_op::store.
```
[the remark this answers is not in the document]
Comment (sarnex) [reply sycl/include/sycl/ext/intel/esimd/memory.hpp:1308]: oh yeah sorry i missed that

File: sycl/include/sycl/ext/intel/esimd/memory.hpp [hunk 7]
```diff
@@ -1213,6 +1213,319 @@ atomic_update(Tx *p, Toffset offset, simd<Tx, N> src0, simd<Tx, N> src1,
...
+                                 simd<Tx, N>>
+    atomic_update(AccessorTy acc, simd<Toffset, N> offset, simd_mask<N> mask) {
+#ifdef __ESIMD_FORCE_STATELESS_MEM
+  return atomic_update<Op, Tx, N>(__ESIMD_DNS::accessorToPointer<Tx>(acc),
+                                  offset, mask);
+#else
+  detail::check_atomic<Op, Tx, N, 0>();
+  static_assert(sizeof(Toffset) == 4, "Only 32 bit offset is supported");
+  if constexpr (Op == atomic_op::load) {
+    return atomic_update<atomic_op::bit_or, Tx, N>(acc, offset, simd<Tx, N>(0),
```
Comment (sarnex) [inline sycl/include/sycl/ext/intel/esimd/memory.hpp:1365]: sorry why is load using bit_or?

File: sycl/include/sycl/ext/intel/esimd/memory.hpp [hunk 8]
```diff
@@ -1213,6 +1213,319 @@ atomic_update(Tx *p, Toffset offset, simd<Tx, N> src0, simd<Tx, N> src1,
...
+                             simd<Tx, N>>
+atomic_update(AccessorTy acc, simd<Toffset, N> offset, simd<Tx, N> src0,
+              simd<Tx, N> src1, simd_mask<N> mask) {
+#ifdef __ESIMD_FORCE_STATELESS_MEM
+  return atomic_update<Op, Tx, N>(__ESIMD_DNS::accessorToPointer<Tx>(acc),
+                                  offset, src0, src1, mask);
+#else
+  detail::check_atomic<Op, Tx, N, 2>();
+  static_assert(sizeof(Toffset) == 4, "Only 32 bit offset is supported");
+  if constexpr (Op == atomic_op::fcmpwr) {
```
Comment (sarnex) [inline sycl/include/sycl/ext/intel/esimd/memory.hpp:1459]: is there some doc explaining the mapping between atomic ops and lsc atomic ops so i can understand why we need to do this if?

**Review by sarnex — APPROVED** [review]
lgtm but we might want slava to review also

File: sycl/include/sycl/ext/intel/experimental/esimd/memory.hpp [hunk 9]
```diff
@@ -2914,15 +2914,17 @@ lsc_atomic_update(T *p, Toffset offset, __ESIMD_NS::simd<T, N> src0,
...
 __ESIMD_API std::enable_if_t<!std::is_pointer<AccessorTy>::value,
                              __ESIMD_NS::simd<T, N>>
-lsc_atomic_update(AccessorTy acc, __ESIMD_NS::simd<uint32_t, N> offsets,
+lsc_atomic_update(AccessorTy acc, __ESIMD_NS::simd<Toffset, N> offsets,
                   __ESIMD_NS::simd_mask<N> pred) {
 #ifdef __ESIMD_FORCE_STATELESS_MEM
   return lsc_atomic_update<Op, T, N, DS, L1H, L3H>(
       __ESIMD_DNS::accessorToPointer<T>(acc), offsets, pred);
 #else
+  static_assert(std::is_integral_v<Toffset> && sizeof(Toffset) == 4,
```
Comment (v-klochkov) [inline sycl/include/sycl/ext/intel/experimental/esimd/memory.hpp:2926]: Generally speaking it is backward incompatibility. It was possible to pass simd<uint64_t,N> Offsets and have automatic conversion at the call-site to simd<uint32_t, N> Offsets.
It may be Ok here taking into account 1) The changes are in experimental 2) It is pretty much a corner case 3) It is a time-windows for api/abi breaking changes now.
@turinevgeny , @sarnex - what do you think?

**Review by v-klochkov — COMMENTED** [review]
Thank you. Please some comments from me.

File: sycl/include/sycl/ext/intel/esimd/memory.hpp (hunk 7 above)
[the remark this answers is not in the document]
Comment (v-klochkov) [reply sycl/include/sycl/ext/intel/esimd/memory.hpp:1365]: > Technically, per check_atomic() we only allow integral types for inc,dec and load.

Do you agree 'load' should not be restricted to integral?
BTW, 'store' is not checked/verified at all in 'check_atomic()'.

File: sycl/include/sycl/ext/intel/experimental/esimd/memory.hpp (hunk 9 above)
Comment (v-klochkov) [reply sycl/include/sycl/ext/intel/experimental/esimd/memory.hpp:2926]: Thank you. I also created a test and reproduced such fail:
```
/include/sycl/ext/intel/experimental/esimd/memory.hpp:2920:1: note: candidate template ignored: could not match 'unsigned int' against 'unsigned long'                                                                                                             
lsc_atomic_update(AccessorTy acc, __ESIMD_NS::simd<uint32_t, N> offsets,
```
It is good then.

File: sycl/include/sycl/ext/intel/esimd/memory.hpp [hunk 10]
```diff
@@ -1069,17 +1070,24 @@ template <atomic_op Op, typename Tx, int N, typename Toffset>
...
   if constexpr (Op == atomic_op::load) {
-    return atomic_update<atomic_op::bit_or, Tx, N>(p, offset, simd<Tx, N>(0),
-                                                   mask);
-  } else {
-    simd<uintptr_t, N> vAddr(reinterpret_cast<uintptr_t>(p));
-    simd<uintptr_t, N> offset_i1 = convert<uintptr_t>(offset);
-    vAddr += offset_i1;
-    using T = typename detail::__raw_t<Tx>;
-    return __esimd_svm_atomic0<Op, T, N>(vAddr.data(), mask.data());
+    if constexpr (std::is_integral_v<Toffset>) {
```
Comment (v-klochkov) [inline sycl/include/sycl/ext/intel/esimd/memory.hpp:1074]: Shouldn't this check be for Tx, not for Toffset?

File: sycl/include/sycl/ext/intel/esimd/memory.hpp [hunk 11]
```diff
@@ -1359,17 +1379,27 @@ __ESIMD_API
...
   }
+
+  detail::check_atomic<Op, Tx, N, 0>();
+  static_assert(sizeof(Toffset) == 4, "Only 32 bit offset is supported");
+
+  static_assert(sizeof(Tx) == 4, "Only 32 bit data is supported");
+  const auto si = __ESIMD_NS::get_surface_index(acc);
+  using T = typename detail::__raw_t<Tx>;
+  return __esimd_dword_atomic0<Op, T, N>(mask.data(), si, offset.data());
+
```
Comment (v-klochkov) [inline sycl/include/sycl/ext/intel/esimd/memory.hpp:1402]: Please put this block to 'else' branch of 'if constexpr (OP == atomic_op::load)' condition.
'if constexpr' works a bit differently than regular 'if', thus this code needs to be in 'else'-branch.

File: sycl/include/sycl/ext/intel/esimd/memory.hpp [hunk 12]
```diff
@@ -1069,17 +1070,24 @@ template <atomic_op Op, typename Tx, int N, typename Toffset>
...
-    simd<uintptr_t, N> offset_i1 = convert<uintptr_t>(offset);
-    vAddr += offset_i1;
-    using T = typename detail::__raw_t<Tx>;
-    return __esimd_svm_atomic0<Op, T, N>(vAddr.data(), mask.data());
+    if constexpr (std::is_integral_v<Toffset>) {
+      return atomic_update<atomic_op::bit_or, Tx, N>(p, offset, simd<Tx, N>(0),
+                                                     mask);
+    } else {
+      using Tint = detail::uint_type_t<sizeof(Tx)>;
+      simd<Tint, N> Res = atomic_update<atomic_op::bit_or, Tint, N>(
```
Comment (v-klochkov) [inline sycl/include/sycl/ext/intel/esimd/memory.hpp:1079]: Thank you for adding support for FP load.
This change requires 2 more fixes:
1) Fix in check_atomic() function - relax restrictions for 'load' - now it is required to be integral.
2) Please add a test for 'load' and 'store' of FP to ensure they really work.

File: sycl/include/sycl/ext/intel/experimental/esimd/memory.hpp [hunk 13]
```diff
@@ -3192,6 +3202,122 @@ atomic_update(T *p, Toffset offset, simd<T, N> src0, simd<T, N> src1,
       p, offset, src1, src0, mask);
 }
 
+template <native::lsc::atomic_op Op, typename T, int N, typename Toffset,
+          typename AccessorTy>
+__ESIMD_API std::enable_if_t<std::is_integral_v<Toffset> &&
+                                 __ESIMD_DNS::get_num_args<Op>() == 0 &&
+                                 !std::is_pointer<AccessorTy>::value,
+                             simd<T, N>>
+atomic_update(AccessorTy acc, simd<Toffset, N> offset, simd_mask<N> mask) {
```
Comment (v-klochkov) [inline sycl/include/sycl/ext/intel/experimental/esimd/memory.hpp:3211]: Why do we have this inconsistency between variants accepting offset as a) simd<Toffset,N>, b) simd_view<>, c) Toffset ?
I.e. why only variants (b) and (c) have the default value for the 'mask' operand?

It looks messy because you can call atomic_update(acc, simd<int,N>{0,4}.select<num,stride>()) without specifying a mask, but removing 'select' and passing offset as a simple simd object would start requiring the mask operand..

Comment (v-klochkov) [reply sycl/include/sycl/ext/intel/experimental/esimd/memory.hpp:3205]: why (c) would cause any problems? It is how to existing lsc_atomic_update() has proven to work, right?

File: sycl/test-e2e/ESIMD/lsc/atomic_smoke.cpp [hunk 14]
```diff
@@ -779,7 +778,6 @@ bool test_int_types_and_sizes(queue q, const Config &cfg) {
   return passed;
 }
 
-#ifndef USE_DWORD_ATOMICS
```
Comment (v-klochkov) [inline sycl/test-e2e/ESIMD/lsc/atomic_smoke.cpp:658]: Why did you remove this macro-guard?
Don't these tests verify fadd/fmin available only on LSC. The called API is old, but it is automatically converted (with a warning) to LSC. Is that correct?

Comment (v-klochkov) [reply sycl/test-e2e/ESIMD/lsc/atomic_smoke.cpp:658]: I see now, the FAdd is still guarded, so, this fix is good

File: sycl/test-e2e/ESIMD/lsc/atomic_smoke.cpp [hunk 15]
```diff
@@ -631,10 +754,10 @@ bool test_fp_types(queue q, const Config &cfg) {
   // passed &= test<sycl::half, N, Op>(q, cfg);
 
   passed &= test<float, N, Op>(q, cfg);
-  passed &= test<double, N, Op>(q, cfg);
+  if (q.get_device().has(sycl::aspect::fp64))
```
Comment (v-klochkov) [inline sycl/test-e2e/ESIMD/lsc/atomic_smoke.cpp:757]: I believe 'aspect-atomic64' needs to be checked here.

File: sycl/include/sycl/ext/intel/experimental/esimd/memory.hpp [hunk 16]
```diff
@@ -2704,7 +2704,7 @@ __ESIMD_API std::enable_if_t<
     __ESIMD_DNS::get_num_args<__ESIMD_DNS::to_lsc_atomic_op<Op>()>() == 0,
     __ESIMD_NS::simd<T, N>>
 lsc_atomic_update(T *p, __ESIMD_NS::simd_view<Toffset, RegionTy> offsets,
-                  __ESIMD_NS::simd_mask<N> pred = 1) {
+                  __ESIMD_NS::simd_mask<N> pred) {
```
Comment (v-klochkov) [inline sycl/include/sycl/ext/intel/experimental/esimd/memory.hpp:2707]: Gregory, why do you change user-visible API so easily/bravely? Such change definitely can cause compilation fails on existing code!

For example, such code compile correctly now (without your fix):
```
            // B is USM pointer 'int *'
            simd<uint32_t, 16> Offsets(0, 8);
            uint32_t OneOffset = 4;
            //lsc_atomic_update<atomic_op::inc, int>(B, Offsets); // Doesn't work only because of inconsistency - lsc_atomic_update accepting simd Offset doesn't have default value for Mask.
            lsc_atomic_update<atomic_op::inc, int, 16>(B, OneOffset); // Compiles correctly.
            lsc_atomic_update<atomic_op::inc, int, 16>(B, Offsets.bit_cast_view<uint32_t>()); // Compiles correctly.
```

**Review by v-klochkov — APPROVED** [review]

**Comment by v-klochkov** [comment]
@fineg74  - 2 tests failed on Windows:
  SYCL :: ESIMD/dword_atomic_smoke.cpp
  SYCL :: ESIMD/dword_atomic_smoke_scalar_off.cpp

**Review by v-klochkov — APPROVED** [review]
